src/els_code_action_provider.erl

```diff
@@ -0,0 +1,183 @@
+    CA1 ++ CA2.
+
+%%------------------------------------------------------------------------------
+
+-spec unused_variable_action(uri(), range(), binary()) -> [map()].
+unused_variable_action(Uri, Range, Message) ->
+    %% processing messages like "variable 'Foo' is unused"
+    %% TODO: this is a hot path, precompile the re's.
+    case re:run(Message, "variable '(.*)' is unused"
+               , [{capture, all_but_first, binary}]) of
+        {match, [UnusedVariable]} ->
+            make_unused_variable_action(Uri, Range, UnusedVariable);
+        _ -> []
+    end.
+
+-spec make_unused_variable_action(uri(), range(), binary()) -> [map()].
+make_unused_variable_action(Uri, Range, UnusedVariable) ->
+    #{ <<"start">> := #{ <<"character">> := _StartCol
+                       , <<"line">>      := StartLine }
+     , <<"end">> := _End
+     } = Range,
+    lager:info("make_unused_variable_action: [Range=~p]"
+              , [Range]),
+    %% processing messages like "variable 'Foo' is unused"
+    lager:info("make_unused_variable_action: [UnusedVariable=~p]"
+              , [UnusedVariable]),
+    Path = binary_to_list(els_uri:path(Uri)),
+    Lines = file2lines(Path),
+    Line = lists:nth(StartLine + 1, Lines),
+    lager:info("make_unused_variable_action: [Line=~p]"
+              , [Line]),
+
+    {ok, Tokens, _} = erl_scan:string(Line, 1, [return, text]),
+    lager:info("make_unused_variable_action: [Tokens=~p]"
+              , [Tokens]),
+    Replace = fun(Tok) ->
+                case Tok of
+                    {var, [{text, VarName}, _], _} -> "_" ++ VarName;
+                    {_,   [{text, Text   }, _], _} -> Text;
+                    {_,   [{text, Text   }, _]}    -> Text
+                end
+              end,
+    UpdatedLine = lists:flatten(lists:map(Replace, Tokens)) ++ "\n",
+    lager:info("make_unused_variable_action: [UpdatedLine=~p]"
+              , [UpdatedLine]),
+    [ replace_lines_action( Uri
+                      , <<"Add '_' to '", UnusedVariable/binary, "'">>
+                      , list_to_binary(UpdatedLine)
+                      , Range)].
+
+%%------------------------------------------------------------------------------
+
+%% Based on https://stackoverflow.com/a/2185533
+-spec file2lines(string()) -> [string()].
+file2lines(File) ->
+   {ok, Bin} = file:read_file(File),
+   string2lines(binary_to_list(Bin), []).
+
+-spec string2lines(string(), string()) -> [string()].
+string2lines("\n" ++ Str, Acc) -> [lists:reverse(Acc) | string2lines(Str, [])];
+string2lines([H|T], Acc)       -> string2lines(T, [H|Acc]);
+string2lines([], Acc)          -> [lists:reverse(Acc)].
+
+%%------------------------------------------------------------------------------
+
+-spec undefined_type_action(uri(), range(), binary()) -> [code_action()].
+undefined_type_action(Uri, Range, Message) ->
+    %% processing messages of the type "type type_name() undefined".
+    %% TODO: this is a hot path, precompile the re's.
```

Comment:
We have a ticket about not having TODOs in the code. Let' s use GitHub issues for follow-up actions.